sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+member functions, which are going to be used from kernels and enforce necessary
+restrictions on them, new compile-time-constant properties are proposed.
+
+[source,dpcpp]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  struct indirectly_callable_key {
+    template <typename SetId>
+    using value_t = property_value<indirectly_callable_key, SetId>;
+  };
+
+  struct calls_indirectly_key {
+    template <typename First = void, typename... SetIds>
+    using value_t = property_value<calls_indirectly_key, First, SetIds...>;
+  };
+
+  template <typename SetId = void>
+  inline constexpr indirectly_callable_key::value_t<SetId> indirectly_callable;
+
+  template <typename First = void, typename... Rest>
+  inline constexpr calls_indirectly_key::value_t<First, Rest...>
+      calls_indirectly;
+
+  template <>
+  struct is_property_key<indirectly_callable_key> : std::true_type {};
+  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
+}
+----
+
+Before describing those properties those properties in more detail, a couple of
+new terms are introduced to simplify the extension specification:
+
+Set of virtual member functions:: a group of virtual member functions which are
+defined with `indirectly_callable` property with the same value of the property
+parameter `SetId`. For the simplicity, this will also be further refferred to as
+a _set_, or as a _set of virtual functions_.
+
+Kernel declares a use of a set of virtual member functions:: a kernel is
+considered to be declaring a use of a set of virtual member functions `SetIdA`
+when it is submitted with `calls_indirectly` property with `SetIdA` included
+into the property parameter `SetIds`. If `SetIdA` is not included into the
+property parameter `SetIds`, or if a kernel is submitted without the property,
+then it is *not* considered to be declaring a use of the set of virtual member
+functions.
+
+|===
+|Property|Description
+|`indirectly_callable`
+|The `indirectly_callable` property indicates that a virtual member function is
+a device function, thus making it available to be called from SYCL kernel and
+device functions. Should only be applied to virtual member functions and to do
+so, function-style `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro should
+be used. The macro accepts a single optional argument, which is passed to the
+property parameter `SetId` and therefore has all the same requirements.
+
+Optional parameter `SetId` specifies a set of virtual member functions this
+function belongs to and at the same time it defines a group of kernels, which
+can call this function, it must be a C++ typename. When the parameter is
+ommitted, a virtual member function is considered to belong to the default set.
+
+Calling a virtual member function from a kernel which does not declare use of a
+set the virtual member function belongs to is an undefined behavior.
+
+|`calls_indirectly`
```

Comment:
Bikeshedding: `calls_indirectly<SetA>` reads well, but I'm worried that `calls_indirectly` alone is confusing.

I'd prefer something like `indirect_calls` (i.e., "this kernel contains indirect calls") or `assume_indirect_calls` (i.e., "the compiler should assume this kernel contains indirect calls").

Reply:
Any other feedback here, @gmlueck maybe?

I'm fine with changing names. We can even have two properties: one that accepts `set` argument and another which doesn't, to avoid the need to add `<>` at the end.

It could be:
```
q.single_task(properties{assume_indirect_calls}, [](){}); // default set
q.single_task(properties{assume_indirect_calls_from<set_A, set_B>}, [](){}); // explicitly-provided sets
```

Reply:
> I presume we would also split `indirectly_callable`?

Right. Those 4 suggested properties looks good to me. @Pennycook, does it resolve your concern? If so, I will update the proposal

Reply:
Yes, I think this is much clearer.

I think the last property should be called something like `indirectly_callable_in`, though.  The symmetry of to/from might imply that the "from" list should be a list of kernels, rather than a list of sets.

Reply:
> I think the last property should be called something like `indirectly_callable_in`, though.

`indirectly_callable_as`? "In" looks a bit weird to me

Reply:
I don't think "as" is quite right, either.  It might help to look at some complete examples.

With "in", I think it's clear from the declarations below  that `foo` is in `set_one` and `set_two`, but `bar` is only in `set_one`.

```c++
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_in<set_one, set_two>)
void foo() {}

virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_in<set_one>)
void bar() {}
```

With "as", the code below looks wrong to me.   Somebody unfamiliar with the extension might read this as setting up aliases: so `foo` is callable via the names `set_one` and `set_two`, whereas `bar` is callable via the name `set_one`.  Such a reader might then be confused about how these two functions can both be called "as" the same thing.
 
```c++
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_as<set_one, set_two>)
void foo() {}

virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_as<set_one>)
void bar() {}
```

I'm not trying to argue that "in" is perfect, and I'd be very happy if we can come up with something better.  But I think "in" is better than "from" and "as".

Reply:
Personally I still prefer "from".  I could live with "in", but I think "as" is confusing.

The words "to" and "from" are a natural pair.  They way I read the properties is:

* `assume_indirect_calls_to<set>` means "assume this kernel might call **to** any function in `<set>`"
* `indirectly_callable_from<set>` means "this function might be called **from** any kernel in `<set>`"

With "in", I guess would read the property as:

* `indirectly_callable_in<set>` means "this function is **in** the set of functions that might be called from kernels in `<set>`.

Reply:
> The words "to" and "from" are a natural pair. They way I read the properties is:
> 
> `assume_indirect_calls_to<set>` means "assume this kernel might call to any function in `<set>`"
> `indirectly_callable_from<set>` means "this function might be called from any kernel in `<set>`"

This is exactly why I don't like "from".  Each  `<set>` contains functions, not kernels -- so what does it mean above for a function to be "called from any kernel in `<set>`"?

With "in" , I think the reading should be:

- `indirectly_callable_in<set>` means "This function is **in** this `<set>` of functions, and may be indirectly called from any kernel that targets this `<set>`".

This matches my understanding of what the extension is doing, at least, but I might have misunderstood something.

Reply:
The way I see it, the "set" applies to both kernels and functions.  We can say that kernels are "in" a set, and we can say that functions are "in" a set.  However, these are both notional concepts.  The implementation doesn't require all functions in the same set to be bundled together into the same device image, for example.  It's just a convenient way to indicate which kernels could make calls to which functions.  The implementation assumes that any kernel in set _A_ might indirectly call any function that is in set _A_.

Reply:
That might work as a solution, but it's not what the specification currently says.  It says:

> For that the pass should compile a list of aspects used by each set of
indirectly callable functions (as defined by `indirectly_callable` property set
by user) and then append those aspects to every kernel which use those sets (as
defined by `calls_indirectly` property set by user).

...which I read as saying that each set is a set of functions, not functions and kernels.

I think if we want to use to/from then the rest of the wording would need to be updated to say explicitly that each set contains a mix of functions and kernels, and that any kernel in a set might indirectly call any function in that set.

Reply:
That quote comes from the design document, not the spec, right?

Reply:
Oops.  Yes, you're right -- sorry, I was searching for "set" and didn't realize there were multiple documents on the page.

But the specification contains similar wording about the set containing only functions:

> Set of virtual member functions:: a group of virtual member functions which are
defined with the `indirectly_callable` property and with the same value of the
property parameter `SetId`. For simplicity, this will also be further referred
to as a _set_, or as a _set of virtual functions_.
> 
> Kernel declares a use of a set of virtual member functions:: a kernel is
considered to be declaring a use of a set of virtual member functions `SetIdA`
when it is submitted with `calls_indirectly` property with `SetIdA` included
into the property parameter `SetIds`. If `SetIdA` is not included into the
property parameter `SetIds`, or if a kernel is submitted without the property,
then it is *not* considered to be declaring a use of the set of virtual member
functions.

Reply:
I think we could adjust the wording in the specification to align with the names we choose.  What is the easiest mental model to understand?  Is it the "to/from" model or the "to/in" model?

Reply:
Sorry for the delayed response.  I'm still leaning towards Alexey's original model of "here are the functions that a kernel can call" -- I think it will be simpler for developers to understand, and it's more aligned with the restrictions.

I still find it weird that a set could contain both functions and kernels.  It just feels a little messy, as abstractions go...

If `kernel_one` contains indirect calls to `foo` and `bar`, it's easy to explain that `kernel_one` includes calls to the sets containing `foo` and `bar`.  Saying `kernel_one` is also a member of those sets feels like a stretch to me.  Even more so if `kernel_two` also contains indirect calls to `foo` and `baz`: that would give us the sets {`kernel_one`, `kernel_two`, `foo`}, {`kernel_one`, `bar`}, {`kernel_two`, `baz`}?  But is there any meaning to the fact that `kernel_one` and `kernel_two` are sometimes in the same set?

If the issue here is really with to/in, maybe we could make the concept of a function set a little more general?

```c++
// Declare that foo is a member of function sets one and two
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::function_set<set_one, set_two>)
void foo() {}

// Declare that this kernel can make indirect calls to sets one and two
q.single_task(properties{assume_indirect_calls_to<set_one, set_two>}, [](){});
```

This avoids having to decide between "from" and "in", but keeps "to" (which we seem to agree has the correct semantics).  As a bonus, we can potentially re-use the concept of a group of device functions for other extensions.

Reply:
> If the issue here is really with from/in, maybe we could make the concept of a function set a little more general?

My concern with this is that the property name `function_set` isn't obviously tied to indirect calls.  For example, people might assume the property guarantees that the functions reside in the same device image (which is not the case).

Reply:
Yeah, I deliberately tried not to tie it to indirection.  It would just be defined as a declared set of functions that other properties could hook into to modify the behavior of all functions in the set.  Maybe that's too far.

Reply:
Yeah, I'd rather not go that far.  We might add sets for some other purpose in the future, and I want to preserve our ability to have "indirect call sets" that are different from "some other sets".

Reply:
Ok, I think we're at an impasse.  I'm still leaning towards to/in as the cleaner mental model.

Should we get a third opinion?

Reply:
I only meant that we should not add the `function_set` property.  I'm OK with the to/in model where the set names are integrated into the "indirectly callable" property.  I think you previously proposed properties like this:

* `assume_indirect_calls`
* `assume_indirect_calls_to<set>`
* `indirectly_callable`
* `indirectly_callable_in<set>`

Refining this a bit, should we allow a pack of set names in `assume_indirect_calls_to`, making it:

* `assume_indirect_calls`
* `assume_indirect_calls_to<set1, set2, ...>`
* `indirectly_callable`
* `indirectly_callable_in<set>`

Reply:
> Refining this a bit, should we allow a pack of set names in `assume_indirect_calls_to`, making it:
> 
> * `assume_indirect_calls`
> * `assume_indirect_calls_to<set1, set2, ...>`
> * `indirectly_callable`
> * `indirectly_callable_in<set>`

This looks good to me. 

Part of me wonders if `indirectly_callable_in` should also accept a pack of set names, so functions can be declared as members of multiple sets.  But I can't immediately think of a use-case, especially if sets are typically going to correspond to required aspects.

Reply:
Renamed the extension doc to reflect new names in b7cdc2d8e86b72368f332fc108faf92b751db205 and updated design doc with them in 46897b3337cadf7c7ba7ed7f66f81ee1e59e9f18